### PR TITLE
DPO

### PR DESCRIPTION
**Pull request by rdyro**
Adding DPO training to the main `train.py` script via config options.

The main new configs are `use_dpo` and other `dpo_...` values. Also `train_data_columns` and 
`eval_data_columns` are now lists of string field `['text']` for LLM pre-training and `['chosen', 'rejected']` for DPO.

Currently, only pre-processed datasets are supported already containing the concatenated full prompts: 'chosen' and 'rejected', like in [https://huggingface.co/datasets/Anthropic/hh-rlhf](https://huggingface.co/datasets/Anthropic/hh-rlhf)

Grain, HF and TFDS data pipelines are all compatible.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

**Comment by rdyro** [comment]
There's a leaked HF token in those commits, but it's been revoked, I'm working on deleting it from history atm.

File: MaxText/train.py
Comment (ZhiyuLi-goog) [inline MaxText/train.py:344]: While the code employs mathematically equivalent calculations, the DPO loss implementation appears slightly different at first glance.
   - here: chosen_logratios = jnp.take_along_axis(chosen_logits - chosen_ref_logits, chosen_ids)
   - [reference](https://github.com/eric-mitchell/direct-preference-optimization/blob/f8b8c0f49dc92a430bae41585f9d467d3618fe2f/trainers.py#L45): chosen_logratios = jnp.take_along_axis(chosen_logits, chosen_ids) - jnp.take_along_axis(chosen_ref_logits, chosen_ref_ids)
  - they are mathematically equivalent but requires some time
 
The separation between `chosen_logratios[..., :-1, :]` and `data["chosen"][..., 1:]` is visually distracting, especially since they're typically grouped together for next-token shifts.

Personally, just curious if it is a better idea to introduce some functions for better modularity and readability only.

Something like:
* get_batch_logps: outputs log_prob after taking in logits and index for both chosen and reject
  - reference: [eric-mitchell dpo](https://github.com/eric-mitchell/direct-preference-optimization/blob/f8b8c0f49dc92a430bae41585f9d467d3618fe2f/trainers.py#L90C6-L90C21) & [mlperf dpo](https://github.com/ZhiyuLi-goog/MoE_study/blob/main/hf/dpo_trainers.py#L50)
* dpo_loss: outputs dpo loss after taking in policy_chosen_logps, policy_rejected_logps,ref_chosen_logps, ref_rejected_logps
  - ref: [eric-mitchell](https://github.com/eric-mitchell/direct-preference-optimization/blob/f8b8c0f49dc92a430bae41585f9d467d3618fe2f/trainers.py#L45) & [mlperf dpo](https://github.com/ZhiyuLi-goog/MoE_study/blob/main/hf/dpo_trainers.py#L8)

**Review by ZhiyuLi-goog — COMMENTED** [review]
@rdyro Thanks so much for all your amazing work on DPO.  We know it wasn't easy and we're almost ready to launch!

**Comment by ZhiyuLi-goog** [comment]
Will double check dataset loading tomorrow.

File: MaxText/input_pipeline/_tfds_data_processing.py
Comment (rdyro) [reply MaxText/input_pipeline/_tfds_data_processing.py:33]: This is to ensure tensorflow does not preallocate GPU memory taking it away from JAX

From here: [https://jax.readthedocs.io/en/latest/gpu_memory_allocation.html#common-causes-of-oom-failures](https://jax.readthedocs.io/en/latest/gpu_memory_allocation.html#common-causes-of-oom-failures)

File: MaxText/train.py
Comment (rdyro) [reply MaxText/train.py:304]: To cast to preferred int type, but it'll probably be better to be explicit `.astype(jnp.int32)`

File: MaxText/train.py
Comment (rdyro) [reply MaxText/train.py:349]: Excellent suggestion, I'll replace this!

File: MaxText/train.py
Comment (rdyro) [reply MaxText/train.py]: Thank you, I was mostly guessing here regarding the correct implementation. I took the gather operation from [https://github.com/huggingface/trl/blob/ed9ea74b62ea5aab3a2c597c4d73fadafc51f5dc/trl/trainer/dpo_trainer.py#L1380](https://github.com/huggingface/trl/blob/ed9ea74b62ea5aab3a2c597c4d73fadafc51f5dc/trl/trainer/dpo_trainer.py#L1380)

`chosen_ref_ids == chosen_ref_ids` - they should be the same right?

I fixed gather (jnp.take_along_axis) to occur twice, it's more readable this way.

When reading DPO reference implementations the helper functions confused me a lot (DPO is quite a simple loss in the end), so I opted for the inlined version, but I'd be happy to modularize them if you think that's better!

File: MaxText/input_pipeline/_tfds_data_processing.py
Comment (ZhiyuLi-goog) [inline MaxText/input_pipeline/_tfds_data_processing.py:34]: nit: do we need to keep these comments?

File: end_to_end/tpu/test_dpo.sh
Comment (ZhiyuLi-goog) [inline end_to_end/tpu/test_dpo.sh:22]: nit: let's remove `max_prefill_predict_length=512` used for inferencing only.

File: MaxText/train.py
Comment (ZhiyuLi-goog) [reply MaxText/train.py]: Thank you @rdyro.

I found a discrepancy between the current implementation and reference one:
* the current implementation: sum(dpo_loss(per_token_logits))
* [reference implementation](https://github.com/eric-mitchell/direct-preference-optimization/blob/main/trainers.py#L45-L52): dpo_loss(sum(per_token_logits))
  - note the shape of `*_logps` is (batch_size,) which has already been reduced on vocab dim in [`get_batch_logps`](https://github.com/eric-mitchell/direct-preference-optimization/blob/main/trainers.py#L90). In other words dpo_loss applies to per each sequence instead of token, aligning the intuition of DPO about preference between different answers (instead of tokens)

At first glance, they might not be equivalent in math, considering some non-linearity inside dpo_loss i.e. softmax.

Given the above discrepancy, I am also wondering if it is slightly better to modularize it for better understanding.

**Comment by ZhiyuLi-goog** [comment]
> Will double check dataset loading tomorrow.

Verified the correctness of dataset loading.

**Comment by ZhiyuLi-goog** [comment]
LGTM once have the above comments addressed. Please resolve and merge.

**Review by ZhiyuLi-goog — APPROVED** [review]

**Comment by peregilk** [comment]
@rdyro Really awesome that DPO is being added to MaxText. Ive been waiting for this a long time. I am on ordinary user that tried giving this branch a spin, and thought I should report my experiences. Not sure if it is relevant since it is not merged yet. In that case, just ignore them.

* For it to run, I needed to specify the axis in train.py. Suspecting this to be related to jax-versions :
` common_prefix_mask = jnp.cumsum(chosen_ids == rejected_ids, axis=-1) == 0`
and
`chosen_logratios = (
      jnp.take_along_axis(chosen_logits[..., :-1, :], chosen_ids[..., None], axis=-1)[..., 0]
      - jnp.take_along_axis(chosen_ref_logits[..., :-1, :], chosen_ids[..., None], axis=-1)[..., 0]
  )
  rejected_logratios = (
      jnp.take_along_axis(rejected_logits[..., :-1, :], rejected_ids[..., None], axis=-1)[..., 0]
      - jnp.take_along_axis(rejected_ref_logits[..., :-1, :], rejected_ids[..., None], axis=-1)[..., 0]
  )` 

* I am getting errors when automatically recovering from checkpoints. I seem to be able to load the final checkpoint when putting this in `load_parameters_path`. However, the starting loss here seems to be the same as the starting loss the last round. So it might seem like that updates are not saved correctly.

* I am getting errors when running evaluation. Trying with the default Anthropic dataset, converted to a hf-dataset (jsonlines) stored in a bucket.

**Comment by peregilk** [comment]
@rdyro I wanted to confirm that I have been able to train a Llama 8B with clear effects from DPO, using this code. The lack of eval makes it easy to overfit, though.

Another issue I am seeing is that setting per_device_batch_size<1 does not work. This means that currently training a 70B model is not possible even on a v5-256.

**Comment by rdyro** [comment]
Thank you @peregilk this is super helpful!

> * For it to run, I needed to specify the axis in train.py. Suspecting this to be related to jax-versions :
>   ` common_prefix_mask = jnp.cumsum(chosen_ids == rejected_ids, axis=-1) == 0`
>   and
>   `chosen_logratios = ( jnp.take_along_axis(chosen_logits[..., :-1, :], chosen_ids[..., None], axis=-1)[..., 0] - jnp.take_along_axis(chosen_ref_logits[..., :-1, :], chosen_ids[..., None], axis=-1)[..., 0] ) rejected_logratios = ( jnp.take_along_axis(rejected_logits[..., :-1, :], rejected_ids[..., None], axis=-1)[..., 0] - jnp.take_along_axis(rejected_ref_logits[..., :-1, :], rejected_ids[..., None], axis=-1)[..., 0] )`

Fixed!

> * I am getting errors when automatically recovering from checkpoints. I seem to be able to load the final checkpoint when putting this in `load_parameters_path`. However, the starting loss here seems to be the same as the starting loss the last round. So it might seem like that updates are not saved correctly.

Fixed!

> * I am getting errors when running evaluation. Trying with the default Anthropic dataset, converted to a hf-dataset (jsonlines) stored in a bucket.

Hmm, I couldn't reproduce it, maybe if you have a small repro, I could take a look or we can follow up on this after this PR merges?

> @rdyro I wanted to confirm that I have been able to train a Llama 8B with clear effects from DPO, using this code. The lack of eval makes it easy to overfit, though.

Fixed, I added `evaluation/dpo_reward_accuracy` metric!

> Another issue I am seeing is that setting per_device_batch_size<1 does not work. This means that currently training a 70B model is not possible even on a v5-256.

I couldn't reproduce this, what error are you getting here?

**Comment by peregilk** [comment]
@rdyro Awesome! I'll test this thoroughly as soon as I am back in the office. Just made a quick test remotely, and got an error because of this setting in dpo.yml

dpo_reference_params_path: 'gs://maxtext-external/logs/dpo_reference_params'

I am not really sure what this is referring to, as this is a private bucket.

**Comment by rdyro** [comment]
> dpo_reference_params_path: 'gs://maxtext-external/logs/dpo_reference_params'
> 
> I am not really sure what this is referring to, as this is a private bucket.
 
This was a WIP way of storing reference params for DPO, I've removed it now since we can use step 0 checkpoint instead

File: MaxText/train.py
Comment (rdyro) [reply MaxText/train.py]: I fully agree with getting the math right, I found anther issue, most implementations add [log_softmax](https://github.com/jcolano/DPO/blob/9ec93f804907179fb66385fc5675fa32a35f22f7/dpo_loss.py#L33) to model outputs before DPO, I added this change.

The discrepancy is there, sorry for that, I fixed it!

On the implementation, I'd really hope to keep the loss in one place, my main difficulty with understanding the DPO loss was how scattered its implementation is across 3 functions, having it in one place should make reading easier. What do you think of this presentation:
```python
# common subsequence and padding mask
common_prefix_mask = jnp.cumsum(chosen_ids != rejected_ids, axis=-1) == 0  # [B, S]
valid_seq_mask = (chosen_segmentation != 0) & (rejected_segmentation != 0) & ~common_prefix_mask  # [B, S]

# compute logratios from the sequence-reduced observed token log-probability
chosen_logps_seq = jnp.take_along_axis(  # [B, S]
    jax.nn.log_softmax(chosen_logits[..., :-1, :], axis=-1), chosen_ids[..., None], axis=-1
)[..., 0]
chosen_logps = jnp.sum(chosen_logps_seq * valid_seq_mask, axis=-1)  # [B]
chosen_ref_logps_seq = jnp.take_along_axis(  # [B, S]
    jax.nn.log_softmax(chosen_ref_logits[..., :-1, :], axis=-1), chosen_ids[..., None], axis=-1
)[..., 0]
chosen_ref_logps = jnp.sum(chosen_ref_logps_seq * valid_seq_mask, axis=-1)  # [B]
chosen_logratios = chosen_logps - chosen_ref_logps  # [B]

rejected_logps_seq = jnp.take_along_axis(  # [B, S]
    jax.nn.log_softmax(rejected_logits[..., :-1, :], axis=-1), rejected_ids[..., None], axis=-1
)[..., 0]
rejected_logps = jnp.sum(rejected_logps_seq * valid_seq_mask, axis=-1)  # [B]
rejected_ref_logps_seq = jnp.take_along_axis(  # [B, S]
    jax.nn.log_softmax(rejected_ref_logits[..., :-1, :], axis=-1), rejected_ids[..., None], axis=-1
)[..., 0]
rejected_ref_logps = jnp.sum(rejected_ref_logps_seq * valid_seq_mask, axis=-1)  # [B]
rejected_logratios = rejected_logps - rejected_ref_logps  # [B]

# DPO loss from chosen and rejected logratios
LABEL_SMOOTHING, BETA = config.dpo_label_smoothing, config.dpo_beta
logratios_delta = BETA * (chosen_logratios - rejected_logratios)  # [B]
loss = (  # [B]
    -jax.nn.log_sigmoid(BETA * logratios_delta) * (1 - LABEL_SMOOTHING)
    - jax.nn.log_sigmoid(-BETA * logratios_delta) * LABEL_SMOOTHING
)
total_loss, total_weights = jnp.sum(loss), jnp.sum(valid_seq_mask)
loss = jnp.sum(loss * (jnp.sum(valid_seq_mask, axis=-1) + 1e-3)) / total_weights
```

**Comment by peregilk** [comment]
> > Another issue I am seeing is that setting per_device_batch_size<1 does not work. This means that currently training a 70B model is not possible even on a v5-256.
> 
> I couldn't reproduce this, what error are you getting here?

@rdyro I am running with a fairly standard setup here with:
`per_device_batch_size=0.5     allow_split_physical_axes=True     ici_tensor_parallelism=8
`

The first error I am facing is trivial. The float is not possible to parse. This is caused by the value in dpo.yml being set to `per_device_batch_size: 2`. The correct setting should be `per_device_batch_size: 2.0`

However, after this I am stuck at:
```
  File "/home/perk/maxtext-env/lib/python3.10/site-packages/absl/app.py", line 254, in _run_main
    sys.exit(main(argv))
  File "/home/perk/maxtext/MaxText/train.py", line 981, in main
    train_loop(config)
  File "/home/perk/maxtext/MaxText/train.py", line 868, in train_loop
    state, metrics = p_train_step(state, example_batch, nextrng)
  File "/home/perk/maxtext/MaxText/train.py", line 528, in train_step
    (loss, aux), raw_grads = grad_func(model, config, data, dropout_rng, state.params, *extra_dpo_args, is_train=True)
  File "/home/perk/maxtext/MaxText/train.py", line 312, in dpo_loss_fn
    data["chosen_segmentation"] = (data["chosen_segmentation"] == 1).astype(jnp.int32)
KeyError: 'chosen_segmentation'
```

For some reason only the following keys are available in data when `per_device_batch_size < 1` at line 312 in train.py:
`dict_keys(['inputs', 'inputs_position', 'inputs_segmentation', 'targets', 'targets_position', 'targets_segmentation'])`

For debugging purposes I did set a breakpoint at the same place when running with `per_device_batch_size = 1`. I then have the following keys available:
`dict_keys(['chosen', 'chosen_position', 'chosen_segmentation', 'rejected', 'rejected_position', 'rejected_segmentation'])`

File: MaxText/checkpointing.py
Comment (ZhiyuLi-goog) [reply MaxText/checkpointing.py:149]: Thank you @rdyro. LGTM!

**Comment by rdyro** [comment]
@peregilk Thanks so much for all the feedback, this is really great!

> The first error I am facing is trivial. The float is not possible to parse. This is caused by the value in dpo.yml being set to `per_device_batch_size: 2`. The correct setting should be `per_device_batch_size: 2.0`

I unfortunately can't reproduce this, do you mind opening a separate issue once we have this DPO branch merged?

> For debugging purposes I did set a breakpoint at the same place when running with `per_device_batch_size = 1`. I then have the following keys available: `dict_keys(['chosen', 'chosen_position', 'chosen_segmentation', 'rejected', 'rejected_position', 'rejected_segmentation'])`

That's good debugging information, this means it's defaulting back to the non-DPO text dataloading, odd, I'm not sure why it'd do that if it has `use_dpo: True`. I couldn't reproduce this either unfortunately, would you mind opening a separate issue after we merge this?

**Comment by peregilk** [comment]
@rdyro My issues with `eval` and `per_device_batch_size<1`, might very well be user errors. I agree it should be moved out of this thread.

I have created [Issue 1089](https://github.com/AI-Hypercomputer/maxtext/issues/1089) that also contains a replication of both errors. Lets continue any discussion under this issue parallel to the merge.

File: MaxText/train.py
Comment (ZhiyuLi-goog) [reply MaxText/train.py]: Sync up and resolved offline!
Thank you @rdyro !

File: MaxText/maxtext_utils.py
Comment (gobbleturk) [inline MaxText/maxtext_utils.py:159]: hero =)

**Review by gobbleturk — APPROVED** [review]
This looks awesome! Thank you! Can we have @aireenmei  review as well

**Review by aireenmei — APPROVED** [review]
LGTM, thanks!

File: MaxText/train.py
Comment (aireenmei) [inline MaxText/train.py:725]: Can we keep this line? tfds and grain pipeline may hang when eval_steps were not set correctly and eval data run out, this line help users track how many eval steps can be completed.

**Review by aireenmei — CHANGES_REQUESTED** [review]

File: MaxText/train.py
Comment (aireenmei) [reply MaxText/train.py:725]: oh, nvm, I see it's moved to later

**Review by aireenmei — APPROVED** [review]

File: MaxText/train.py
Comment (rdyro) [reply MaxText/train.py:725]: But it was out of order with the iterator increment, so it needed a fix, thanks for catching this!

**Pull request merged**
